Clean up work on constants.

YashJipkate opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
This PR is a clone of #6915.
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

YashJipkate committed:
Applied stash

vinitamurthi reviewed (approved):
Approving for question-player and practice-session files

YashJipkate commented:
@seanlip @kevinlee12 The setup test shows pending here while it has passed as I saw in the dashboard:
![Screenshot from 2019-06-14 07-40-29](https://user-images.githubusercontent.com/34203227/59478719-e70aae00-8e77-11e9-84a3-28334bd53a7e.png)

Also, the frontend and backend tests are passing which implies that the setup test must have passed.

kevinlee12 commented on core/templates/dev/head/pages/topic-editor-page/topic-editor-page.constants.ts:
```diff
@@ -0,0 +1,30 @@
+// Copyright 2018 The Oppia Authors. All Rights Reserved.
```
Year :)

seanlip replied:
FYI, don't worry about year. I think we should either keep it in sync everywhere or not worry about it.

kevinlee12 reviewed (approved):
lgtm!

I think we can probably override it since this is because Github isn't reporting the correct status.

seanlip commented:
No, we need to fix it so that GitHub reports the correct status. I am not
overriding without all checks passing.

seanlip commented:
Oh, that said, I am not too concerned about "setup" if the others do pass. (I made the original response via email.)

/cc @oppia/dev-workflow-team any idea why this is happening?

YashJipkate commented:
@seanlip The other tests do pass! Could this be merged then?

seanlip commented on core/templates/dev/head/app.constants.ts:
```diff
@@ -0,0 +1,163 @@
+ * @fileoverview Shared constants for the Oppia module.
+ */
+
+oppia.constant('RULE_SUMMARY_WRAP_CHARACTER_COUNT', 30);
```
Just to confirm ... you've verified that every single constant in this file is used in > 1 file? If not, please do.

YashJipkate replied:
@seanlip I previously followed the way that all the constants in App.ts are to be in the app level file. But there were some which were not used in multiple files. So I have put them in their respective page constants file. PTAL!

YashJipkate replied:
Also there were some constants used in single file but they were from the services/ folder which is a shared directory and hence such constants are still in this app level file.

EDIT: Alternatively, we can transfer such constants to the respective files of their usage in services/. But then there would be no *.constants.ts file in services/ directory since that convention is not followed there.

What should be done? I prefer the second one.

seanlip replied:
I don't mind as long as we have a clear and consistent convention that developers can easily understand.

seanlip replied:
I also suggest that you include, with this PR, a means of enforcing this convention through the lint checks.

YashJipkate replied:
@seanlip I have added the check! PTAL!

seanlip replied:
It's a good check, but I think you also need a check to ensure that each constant is declared exactly once globally.

seanlip replied:
@YashJipkate Please respond to all inline comments.

YashJipkate replied:
Sorry for not responding. But this comment is already addressed. I have incorporated a lint check to ensure no repetition of constant declaration.

seanlip commented on core/templates/dev/head/components/question-directives/question-editor/question-editor.constants.ts:
```diff
@@ -0,0 +1,19 @@
+ * @fileoverview Constants for the questions editor directive.
+ */
+
+oppia.constant('INTERACTION_SPECS', GLOBALS.INTERACTION_SPECS);
```
If INTERACTION_SPECS is used in both question-editor and question-player (and possibly elsewhere), shouldn't it be an app-level constant? Or perhaps not at the root, but somewhere shared? We should not be duplicating code between question-editor and question-player (or multiple files in general).

Also, what steps are you planning to take (in terms of presubmit checks) to ensure that every constant is declared in at most one place? That's something that can definitely happen if there are constants in multiple files.

YashJipkate replied:
Actually, this is a special type of "constant". It uses the GLOBALS variable which is different for every page and hence needs to be declared at individual pages. I tried putting it at the and it expectedly gave errors.

seanlip replied:
I think your last sentence is missing a word :)

Also I think @vojtechjelinek and @Jamesjay4199 might need to comment on this, this sounds like something that needs to be fixed (and not in this way).

vojtechjelinek replied:
If you look at https://github.com/oppia/oppia/search?l=Python&q=INTERACTION_SPECS you can see that `INTERACTION_SPECS` is always set to `interaction_registry.Registry.get_all_specs()` so the value is always the same, the problem is that we are sending this only on some pages and if we were to move to app-level constant we would have to send it with every page.

Let's keep multiple constants for now and @Jamesjay4199 will refactor this in his third milestone.

seanlip replied:
OK, two follow-up questions then.

(1) 

> It uses the GLOBALS variable which is different for every page and hence needs to be declared at individual pages. I tried putting it at the and it expectedly gave errors.

@YashJipkate can you elaborate on this (the "different for every page" part)? It contradicts with what @vojtechjelinek said, so I would like to understand the evidence for your statement.

(2) @vojtechjelinek I don't think we can support multiple repetitions of the same constant. Can we have a separate file containing just this constant that is imported by the different pages? That way the constant stays in one place, and @YashJipkate can put in checks to enforce the constants policy. There should be a TODO and an issue to delete this file and that issue should be assigned to @Jamesjay4199. Would that work?

YashJipkate replied:
@seanlip I did not know the source of the variable (that it was coming from backend) and since it was giving out errors, so I thought maybe the constant was not a constant for every page. You can ignore that statement of mine.

vojtechjelinek replied:
@seanlip (2) Sure that's okay

jameesjohn replied:
Yeah, (2) seems okay @seanlip

seanlip commented on core/templates/dev/head/pages/exploration-editor-page/exploration-editor-page.constants.ts:
```diff
@@ -67,3 +52,5 @@ oppia.constant(
 
 oppia.constant('DEFAULT_AUDIO_LANGUAGE', 'en');
+
+oppia.constant('INTERACTION_SPECS', GLOBALS.INTERACTION_SPECS);
```
Hm, a third repetition?

seanlip commented on core/templates/dev/head/pages/exploration-player-page/exploration-player-page.constants.ts:
```diff
@@ -52,23 +42,56 @@ oppia.constant(
+  STATS_EVENTS: '/explorehandler/stats_events/<exploration_id>'
+});
+
+oppia.constant('INTERACTION_SPECS', GLOBALS.INTERACTION_SPECS);
```
And a fourth

seanlip commented on core/templates/dev/head/pages/practice-session-page/practice-session-page.constants.ts:
```diff
@@ -18,8 +18,8 @@
   'PRACTICE_SESSIONS_DATA_URL',
   '/practice_session/data/<topic_name>');
+
+oppia.constant('INTERACTION_SPECS', GLOBALS.INTERACTION_SPECS);
```
And a fifth

seanlip reviewed (changes requested):
Hi @YashJipkate, took a pass, and I have a question. I think each constant should be declared exactly once in all of Oppia. I'm fine having a policy of "if a constant is used more than once, put it at app level", though there may be other better ways of handling this. 

What are the policies that you would like to put in place, in order to achieve this? I would also like the presubmit checks that force this policy to be followed in the future to be part of this PR, so that devs don't (accidentally) undo the work you've done to separate them properly.

seanlip commented:
@vojtechjelinek @Jamesjay4199 Flagging [this comment](https://github.com/oppia/oppia/pull/6919#discussion_r293706538), PTAL.

YashJipkate committed:
Moved single use constants to page.constants

YashJipkate committed:
Fix e2e by requiring the necessary constants file

YashJipkate committed:
Moved single use constants to services/ where they belong

YashJipkate committed:
Added presubmit check

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -2414,6 +2419,50 @@ def _check_extra_js_files(self):
 
         return summary_messages
 
+    def _check_constants_declaration(self):
```
This is a good check, but I thought the point of the comment thread was to have a check to ensure that no constant is declared more than once in the codebase?

seanlip replied:
@YashJipkate please respond to all inline comments.

seanlip replied:
@YashJipkate Repeat: Please respond to all inline comments.

YashJipkate replied:
This concern is already addressed in the later commits. This somehow missed my radar.

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -340,6 +340,11 @@
     '/scripts/install_third_party.sh',
     '/.github/']
 
+NEW_CONVENTION_DIRECTORIES = [
```
Instead of this, please use OLD_CONVENTION_DIRECTORIES and check "all frontend files minus stuff in OLD_CONVENTION_DIRECTORIES". Then, gradually reduce the list over time until we can remove it altogether.

YashJipkate replied:
Done!

seanlip reviewed (commented):
Two comments @YashJipkate (the one above and the one below). PTAL.

YashJipkate committed:
Merge branch 'develop' into cl-sep-co

YashJipkate committed:
Added presubmit check to ensure singular declaration of constants

YashJipkate committed:
Fix lint

YashJipkate commented:
@seanlip PTAL!

seanlip commented:
@YashJipkate did you see this comment? https://github.com/oppia/oppia/pull/6919#discussion_r293936636

seanlip commented:
@YashJipkate I took a look, but I am not going to approve this until the INTERACTION_SPECS constant appears in only one place. You should have a backend lint check for this. Please read the comment history carefully and make sure that you address all open comments before requesting a review.

YashJipkate committed:
Merge branch 'develop' into cl-sep-co

YashJipkate committed:
Removed individual declaration of INTERACTION_SPECS

YashJipkate commented:
> @YashJipkate did you see this comment? [#6919 (comment)](https://github.com/oppia/oppia/pull/6919#discussion_r293936636)

I have already addressed that comment, Sorry for not saying the same on the thread.

> but I am not going to approve this until the INTERACTION_SPECS constant appears in only one place.

I thought that you meant this was something @Jamesjay4199 would have as a TODO. MIght be some misunderstanding of mine.
I have made a new file which has the constant and imported it wherever needed. PTAL!

YashJipkate commented on .github/CODEOWNERS:
```diff
@@ -314,6 +314,8 @@
 /core/platform/ @seanlip
 /core/templates/dev/head/App*.ts @ankita240796
+/core/templates/dev/head/app.constants.ts @ankita240796
+/core/templates/dev/head/pages/interaction-specs.contants.ts @ankita240796
```
@seanlip I am not sure about this. Could you confirm?

seanlip replied:
I would say @vojtechjelinek (what do you think, @vojtechjelinek)?

YashJipkate commented on scripts/pre_commit_linter.py:
```diff
@@ -2421,6 +2430,77 @@ def _check_extra_js_files(self):
+                                   'separate constants file.' % (
+                                       filepath, line_num))
+
+                # Check if the constant has multiple declarations which is
```
This is the part where it checks for multiple declarations.

seanlip replied:
Make sure to update the function docstring please.

YashJipkate replied:
Done.

seanlip commented:
> I have already addressed that comment, Sorry for not saying the same on the thread.

But shouldn't that check have caught the case of INTERACTION_SPECS being repeated in the codebase, if it was working correctly?

YashJipkate commented:
@seanlip The tests now pass ([on my local](https://travis-ci.com/YashJipkate/oppia/builds/115700783)). PTAL!

YashJipkate commented:
> But shouldn't that check have caught the case of interaction-specs being repeated in the codebase, if it was working correctly?

I had an exception for INTERACTION_SPECS in the linter so it is exempted.

YashJipkate commented on scripts/pre_commit_linter.py:
```diff
@@ -2451,6 +2457,32 @@ def _check_constants_declaration(self):
+                    else:
+                        constant_name = expression.arguments[0].raw
+                        if constant_name == '\'INTERACTION_SPECS\'':
+                            continue
```
This was where I had laid out an exemption case for INTERACTION_SPECS.

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -347,6 +347,15 @@
 
+OLD_CONVENTION_DIRECTORIES = [
+    'core/templates/dev/head/services',
+    'core/templates/dev/head/base_components',
```
Isn't this new? It wasn't around in the "old" codebase...

YashJipkate replied:
I don't think I understood you completely. These are the directories which follow the old convention and hence the name.

seanlip replied:
I guess I was thinking that base_components seems to have been recently added and is here to stay? I thought the idea was that all "new" dirs are cleaned up and we were going to remove all the dirs in this list completely over time... (which, btw, merits an issue or a TODO).

Are we planning to delete base_components too?

YashJipkate replied:
No, it doesn't need to be deleted. It just means that it is following the old convention. When we will empty this list out, it would mean that these directories now follow the new convention.

seanlip replied:
OK. I think you missed adding both an issue and a TODO for this (see previous comment). We need some way of keeping track of the task of "making this list empty".

Also add a comment above line 350 to explain to new devs what this constant is (in particular what is meant by "old convention").

YashJipkate replied:
I have added a comment as well a TODO. PTAL.

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -2421,6 +2430,77 @@ def _check_extra_js_files(self):
+                    filepath.endswith('.ts'))]
+            ts_files_to_check = [
+                filepath for filepath in all_ts_files if not any(
+                    pattern in filepath for pattern in (
```
I think this should be stricter: if not filepath.startswith(pattern)

YashJipkate replied:
Done. Replaced with `filepath.startswith(pattern)`

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -2421,6 +2430,77 @@ def _check_extra_js_files(self):
+                filepath for filepath in all_ts_files if not any(
+                    pattern in filepath for pattern in (
+                        OLD_CONVENTION_DIRECTORIES)) and filepath.endswith(
+                            'App.ts')]
```
Not great to have this special case. Make it part of OLD_CONVENTION_DIRECTORIES?

YashJipkate replied:
But then it is not a directory. It is not suited in the list. We would have to make a different list which would just have this file. Or change the name from OLD_CONVENTION_DIRECTORIES to OLD_CONVENTION_PATHS?

seanlip replied:
Paths is fine.

YashJipkate replied:
Done.

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -2421,6 +2430,77 @@ def _check_extra_js_files(self):
+                    pattern in filepath for pattern in (
+                        OLD_CONVENTION_DIRECTORIES)) and filepath.endswith(
+                            'App.ts')]
+            constants_number_dict = {}
```
constants_to_source_filepaths

(you're not storing any numbers here, AFAIK)

YashJipkate replied:
Done!

YashJipkate committed:
Test

YashJipkate committed:
Address Review

YashJipkate committed:
Fix tests

YashJipkate committed:
Merge branch 'develop' into cl-sep-co

seanlip commented on .github/CODEOWNERS:
```diff
@@ -315,7 +315,7 @@
 /core/templates/dev/head/App*.ts @ankita240796
 /core/templates/dev/head/app.constants.ts @ankita240796
-/core/templates/dev/head/pages/interaction-specs.contants.ts @ankita240796
+/core/templates/dev/head/pages/interaction-specs.constants.ts @ankita240796
```
Change to @vojtechjelinek in a future PR since @Jamesjay4199 is working on a refactor later I think.

YashJipkate replied:
Done!

seanlip reviewed (approved):
Basically LGTM. Thanks @YashJipkate! 

Is this PR blocking future work? If so, then if Travis tests pass you are welcome to incorporate the changes from here in the next PR instead, so that we can get this in.

YashJipkate committed:
Address review

YashJipkate commented:
Hi @seanlip. No, this is not blocking any of my work currently (the one actually blocking is #6939) since work on Angular migration on constants can only begin after hybridization which is blocked by #6939. This PR is just a preparation for the next task after hybridization.

So I am fine addressing your review comments in this PR itself!
Thanks!

seanlip commented:
Thanks @YashJipkate! Full LGTM from me. But please file an issue for  https://github.com/oppia/oppia/pull/6919#discussion_r294097847, assign it to yourself, and plan to take care of it as part of your project.

seanlip commented:
Also since this change is mechanical I am happy to merge it once the Travis checks pass.

YashJipkate commented:
@seanlip we already have an issue regarding separation of constants. See #6731. This PR fixes a part of that issue. So may be we can just update its description to apply this to the remaining folders?

seanlip commented:
Sure, that's fine, as long as it's tracked. Are you planning to take that up?

(I don't want that list of folders to stay in the linter script as a long-term thing...)

YashJipkate commented:
OK @seanlip I will try to do that right away in a follow up PR since that needs to be done sooner or later!

seanlip commented:
Thanks, @YashJipkate! I will merge this one now.

YashJipkate commented:
Also the tests pass. Could this be merged?

Edit: Thanks for merging!
